File: cases/taint/kotlin/deep-relay-chain-negative/DeepRelayChainNegative.kt
package dataflowbench

object DeepRelayChainNegative {
    fun dfb_source(): String { // DFB-SOURCE: deep-relay-chain-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: deep-relay-chain-sink

    fun relay1(value: String): String { // DFB-WITNESS: deep-relay-chain-hop1
        return relay2(value)
    }

    fun relay2(value: String): String { // DFB-WITNESS: deep-relay-chain-hop2
        return relay3(value)
    }

    fun relay3(value: String): String { // DFB-WITNESS: deep-relay-chain-hop3
        return relay4(value)
    }

    fun relay4(value: String): String { // DFB-WITNESS: deep-relay-chain-hop4
        return relay5(value)
    }

    fun relay5(value: String): String { // DFB-WITNESS: deep-relay-chain-hop5
        return relay6(value)
    }

    fun relay6(value: String): String { // DFB-WITNESS: deep-relay-chain-hop6
        return value
    }

    fun run() {
        val tainted = dfb_source()
        dfb_sink(relay1("clean"))
    }
}

File: cases/taint/c/deep-relay-chain-negative/deep_relay_chain_negative.c
int dfb_source(void) { // DFB-SOURCE: deep-relay-chain-input
    return 1;
}

int relay6(int value) { // DFB-WITNESS: deep-relay-chain-hop6
    return value;
}

int relay5(int value) { // DFB-WITNESS: deep-relay-chain-hop5
    return relay6(value);
}

int relay4(int value) { // DFB-WITNESS: deep-relay-chain-hop4
    return relay5(value);
}

int relay3(int value) { // DFB-WITNESS: deep-relay-chain-hop3
    return relay4(value);
}

int relay2(int value) { // DFB-WITNESS: deep-relay-chain-hop2
    return relay3(value);
}

int relay1(int value) { // DFB-WITNESS: deep-relay-chain-hop1
    return relay2(value);
}

void dfb_sink(int value) {} // DFB-SINK: deep-relay-chain-sink

void run(void) {
    int tainted = dfb_source();
    dfb_sink(relay1(0));
}
